cranelift/wasm/src/code_translator.rs

```diff
@@ -2542,6 +2574,91 @@ pub fn optionally_bitcast_vector(
     }
 }
 
+#[inline(always)]
+fn is_non_canonical_v128(ty: ir::Type) -> bool {
+    match ty {
+        B8X16 | B16X8 | B32X4 | B64X2 | I64X2 | I32X4 | I16X8 | F32X4 | F64X2 => true,
+        _ => false,
+    }
+}
+
+/// Cast to I8X16, any vector values in `values` that are of "non-canonical" type (meaning, not
+/// I8X16) and park them in the supplied smallvec `canonicalised`.  But before doing that, make
+/// a pre-scan to check if any of this is necessary, and return `false` if not.
+fn canonicalise_v128_values(
+    canonicalised: &mut SmallVec<[ir::Value; 16]>,
+    builder: &mut FunctionBuilder,
+    values: &[ir::Value],
+) -> bool {
```

Comment:
If we return `Option<SmallVec::<[ir::Value; 16]>>` here, we don't have to do `SmallVec::<[ir::Value; 16]>::new()` logic in the `canonicalise_then_`

Reply:
The reason for taking a `&mut SmallVec<[ir::Value; 16]>` rather than returning a `Option<SmallVec::<[ir::Value; 16]>>` is to avoid copying `Option<SmallVec::<[ir::Value; 16]>>` in the return at each call.  In the way I had it, `canonicalise_v128_values` never has to copy the vector (at return), because the caller owns it, and it very rarely even has to put anything in the vector, because most `values` are all non-vector.

Reply:
One simplification that could also make this more convenient to call would be the signature:

```rust
fn canonicalize_v128_values<'a>(
    canonicalized: &'a mut SmallVec<[ir::Value; 16]>,
    builder: ..,
    values: &'a [ir::Value],
) -> &'a [ir::Value] {
    // ...
}
```

which either returns the original `values` argument or returns a slice of the `canonicalised` argument.

Reply:
@alexcrichton Ah, that's neat.  I'll do that.